src/Microsoft.DotNet.ProjectJsonMigration/transforms/TransformApplicator.cs

```diff
@@ -0,0 +1,170 @@
+                    var mergeResult = MergeItems(item, existingItem);
+                    item = mergeResult.InputItem;
+
+                    // Existing Item is null when it's entire set of includes has been merged with the MergeItem
+                    if (mergeResult.ExistingItem == null)
+                    {
+                        existingItem.Parent.RemoveChild(existingItem);
+                    }
+
+                    Execute(mergeResult.MergedItem, destinationItemGroup);
+                }
+
+                // Item will be null only when it's entire set of includes is merged with existing items
+                if (item != null)
+                {
+                    Execute(item, destinationItemGroup);
+                }
+            }
+            else
+            {
+                Execute(item, destinationItemGroup);
+            }
+        }
+
+        public void Execute(
+            IEnumerable<ProjectItemElement> items,
+            ProjectItemGroupElement destinationItemGroup,
+            bool mergeExisting)
+        {
+            foreach (var item in items)
+            {
+                Execute(item, destinationItemGroup, mergeExisting);
+            }
+        }
+
+        /// <summary>
+        /// Merges two items on their common sets of includes.
+        /// The output is 3 items, the 2 input items and the merged items. If the common
+        /// set of includes spans the entirety of the includes of either of the 2 input
+        /// items, that item will be returned as null.
+        ///
+        /// The 3rd output item, the merged item, will have the Union of the excludes and
+        /// metadata from the 2 input items. If any metadata between the 2 input items is different,
+        /// this will throw.
+        ///
+        /// This function will mutate the Include property of the 2 input items, removing the common subset.
+        /// </summary>
+        private MergeResult MergeItems(ProjectItemElement item, ProjectItemElement existingItem)
+        {
+            if (!string.Equals(item.ItemType, existingItem.ItemType, StringComparison.Ordinal))
+            {
+                throw new InvalidOperationException("Cannot merge items of different types.");
+            }
+
+            if (!item.CommonIncludes(existingItem).Any())
+            {
+                throw new InvalidOperationException("Cannot merge items without a common include.");
+            }
+
+            var commonIncludes = item.CommonIncludes(existingItem).ToList();
+            item.RemoveIncludes(commonIncludes);
+            existingItem.RemoveIncludes(commonIncludes);
+
+            var mergedItem = _projectElementGenerator.AddItem(item.ItemType, string.Join(";", commonIncludes));
+
+            mergedItem.AddExcludes(existingItem.Excludes());
+            mergedItem.AddExcludes(item.Excludes());
+
+            mergedItem.AddMetadata(existingItem.Metadata);
+            mergedItem.AddMetadata(item.Metadata);
```

Comment:
You understand the system I wish we had instead of the one we actually do have. In C# terms it's basically

```C#
class StringWithMetadata {
   public string IncludeString;
   public Dictionary<string, string> Metadata;
}
class Item : List<StringWithMetadata> {}
```

Each XML element inside an `ItemGroup` essentially does `item.Append(new StringWithMetadata(metadataXML))`.

So in my made-up scenario, there are two `X` items with the include value `x`, and they have separate metadata.

This is super confusing! I thought for years that items were sets instead of lists.

But this is all a bit off in the weeds: I think it's fine to do coalescing like this in your case, because you know you won't be depending on having the same thing in a list multiple times. None of the "standard" item types do that I know of.